### ISSUE TITLE
Refactor speech to support callbacks, speech commands in control/format fields, priority output, etc.

### ISSUE THREAD

**Issue by nvaccessAuto** [nvaccess/nvda#4877]
**Reported by jteh on 2015-02-03 00:56**
A lot of potentially nice functionality is not possible (or is at least ridiculously painful) with our current speech framework. Speech needs a pretty big refactor to allow for such things. It needs to support:
- Callbacks which are called at a requested point in the speech output
- Callbacks which are called when a given synth finishes speaking and when overall speech has stopped
- Speech commands in control/format field speech so that control and formatting info can be indicated by things other than just text
- Priority output so that important messages can interrupt what is being spoken and/or be spoken after the next utterance without losing lower priority utterances already sent

Unfortunately, some of this is going to break backwards compatibility, but I think it's worth it in this case.
**Blocking #905, #3188, #3286, #3736, #4089, #4233, #4433, #4874, #4966, #5026, #5104**

**Comment by camlorn** [comment]
I thought of another possible use case, though I'm not sure how useful it would be: implement an option to monitor things in the background.  I'm thinking at least aria live regions and the controller API, but it might also be useful to tag terminals for watching.
As an example, I'm currently programming something wherein I need to have two terminals open, one running a client and the other running a server.  If I could tag them both with separate voices and monitor even when it doesn't have focus, that might be very useful.  It could also be horribly confusing, mind you, and it would probably help if this also included pan.  But it's an interesting thought and maybe something to put on the "We want an add-on to be able to..." list at least.  I don't think you could use it instantly, but I think you could train to it if you tried a bit.
There is an old demo somewhere of a prototype system that did something like this for other stuff as well, speaking labels out one speaker and control types out the other in parallel. I wish I had a link to it.  I'll see if I can find it, but don't even know what to search for.
But we're still collecting use cases, so I figured I'd throw this out there anyway.  I'm sure it has problems I haven't thought of, etc, but it's something to put on the list.

**Comment by camlorn** [comment]
Got one more.

On the web, we currently don't read the title attribute when moving by arrows.  I think this also applies to aria-label and maybe some other things.  With this, it would be possible to indicate these attributes efficiently, so that a user could know to check it in one way or another.

As it stands, we would have to say something like "has title" or just read it every time.  Since this can be used on things like abbreviations, this is possibly less than useful.

There's already an issue about them not reading, but I don't remember which one at the moment.  Just that it has a weird title and that I'm on the thread.

**Comment by amangano-edx** [comment]
I posted on bugzilla about an issue with NVDA (and JAWS) on Firefox where the 'Remember Password' popup interrupts other aria alerts on the page (https://bugzilla.mozilla.org/show_bug.cgi?id=1323070). It was determined that this was a problem with the screen readers not being able to handle multiple alerts at the same time. Would the changes proposed here fix this issue?

**Comment by camlorn** [comment]
It wouldn't, I don't think.  It would allow cool experimental things like saying them all at the same time in different voices and maybe panning them across the sound field or something, but the core issue is probably about queuing things to be spoken, and we should already be able to do that.

**Comment by jcsteh** [comment]
Actually, we can't currently handle queuing speech when there are multiple
interrupting messages. How to handle this well is somewhat of an open
question, since many things intentionally interrupt speech multiple times
when there are several updates in quick succession and we really don't want
to queue all of that now irrelevant junk. (See #6335, where the author
intentionally does this with an assertive live region.)

**Comment by leonardder** [comment]
An additional use case I'd like to suggest is an UI to reorder speech output. E.g. 

1. Save configuration on exit  check box  checked  Alt+s
2. checked Save configuration on exit  check box Alt+s
3. checked check box Save configuration on exit  Alt+s
4. check box checked Save configuration on exit  Alt+s

Such an ui could have a list containing the several speech parts, along with move up and move down buttons. This is especially helpful in cases where one wants to reorder output foor table cells in Excel (e.g. speak column and row headers before the cell contents).

**Comment by jcsteh** [comment]
I understand the request, but that's a use case that isn't really covered
by the kind of work that is being planned here. There's probably an
existing issue for this, but if there isn't, feel free to open one.

**Comment by bhavyashah** [comment]
List of tickets which may be related to, blocked by  or dependent on this ticket:
https://github.com/nvaccess/nvda/issues/4966
https://github.com/nvaccess/nvda/issues/5026
https://github.com/nvaccess/nvda/issues/5096
https://github.com/nvaccess/nvda/issues/5104
https://github.com/nvaccess/nvda/issues/3286
https://github.com/nvaccess/nvda/issues/5638
https://github.com/nvaccess/nvda/issues/5862
https://github.com/nvaccess/nvda/issues/1229
https://github.com/nvaccess/nvda/issues/6360
https://github.com/nvaccess/nvda/issues/3564
https://github.com/nvaccess/nvda/issues/4738
https://github.com/nvaccess/nvda/issues/6685
https://github.com/nvaccess/nvda/issues/6688
https://github.com/NVDARemote/NVDARemote/issues/110
https://github.com/nvaccess/nvda/issues/4433
https://github.com/nvaccess/nvda/issues/4629
https://github.com/nvaccess/nvda/issues/310
https://github.com/nvaccess/nvda/issues/7274
https://github.com/nvaccess/nvda/issues/3772
https://github.com/nvaccess/nvda/issues/3493
https://github.com/nvaccess/nvda/issues/4874
https://github.com/nvaccess/nvda/issues/4661
https://github.com/nvaccess/nvda/issues/2590
https://github.com/nvaccess/nvda/issues/2670
https://github.com/nvaccess/nvda/issues/3807
https://github.com/nvaccess/nvda/issues/279
https://github.com/nvaccess/nvda/issues/905
https://github.com/nvaccess/nvda/issues/847
https://github.com/nvaccess/nvda/issues/7427
https://github.com/nvaccess/nvda/issues/3188
https://github.com/nvaccess/nvda/issues/3736
https://github.com/nvaccess/nvda/issues/4086
https://github.com/nvaccess/nvda/pull/7594
https://github.com/nvaccess/nvda/issues/1398
https://github.com/nvaccess/nvda/issues/4089
https://github.com/nvaccess/nvda/issues/4233

**Comment by Adriani90** [comment]
@michaelDCurran I propose to let this issue open because here are many issues referenced which contain use cases for this feature. After solving and closing them, I propose to close this as well.

**Comment by leonardder** [comment]
I've looked at the initial description of this issue, and I think it is fully covered by #4877. Leaving this issue open until all the mentioned issues are solved as @adriani90 suggested, will probably mean that this will be open for ever. Furthermore, there's nothing holding you back from referring to this issue to find the use cases for the framework grouped together.

Closing it.
